barman/recovery_executor.py

```diff
@@ -1238,12 +1224,299 @@ def close(self):
         """
         # Remove the temporary directories
         for temp_dir in self.temp_dirs:
-            shutil.rmtree(temp_dir, ignore_errors=True)
+            temp_dir.delete()
         self.temp_dirs = []
 
-    def _pg_config_mangle(
-        self, filename, settings, backup_filename=None, append_lines=None
+
+class TarballRecoveryExecutor(RecoveryExecutor):
+    """
+    A specialised recovery method for compressed backups.
+    Inheritence is not necessarily the best thing here since the two RecoveryExecutor
+    classes only differ by this one method, and the same will be true for future
+    RecoveryExecutors (i.e. ones which handle encryption).
+    Nevertheless for a wip "make it work" effort this will do.
+    """
+
+    BASE_TARBALL_NAME = "base"
+
+    def __init__(self, backup_manager, compression):
+        """
+        Constructor
+
+        :param barman.backup.BackupManager backup_manager: the BackupManager
+            owner of the executor
+        :param compression Compression.
+        """
+        super(TarballRecoveryExecutor, self).__init__(backup_manager)
+        self.compression = compression
+
+    def _backup_copy(
+        self,
+        backup_info,
+        dest,
+        tablespaces=None,
+        remote_command=None,
+        safe_horizon=None,
+        recovery_info=None,
+    ):
+        # Set a ':' prefix to remote destinations
+        dest_prefix = ""
+        if remote_command:
+            dest_prefix = ":"
+
+        # Instead of adding the `data` directory and `tablespaces` to a copy
+        # controller we instead want to copy just the tarballs to a staging
+        # location via the copy controller and then untar into place.
+
+        # Create the staging area
+        staging_dir = os.path.join(
+            self.config.recovery_staging_path,
+            "barman-staging-{}-{}".format(self.config.name, backup_info.backup_id),
+        )
+        output.info(
+            "Staging compressed backup files on the recovery host in: %s", staging_dir
+        )
+        recovery_info["cmd"].create_dir_if_not_exists(staging_dir, mode="700")
+        recovery_info["cmd"].validate_file_mode(staging_dir, mode="700")
+        recovery_info["staging_dir"] = staging_dir
+        self.temp_dirs.append(
+            fs.UnixCommandPathDeletionCommand(staging_dir, recovery_info["cmd"])
```

Comment:
Using `temp_dirs` to ensure the staging dir is cleaned up means it's going to get cleaned up whenever `close()` is called, so we won't keep the staging dir around in the event of failure.

Maybe that's ok for now, in which case we'd just need to update the documentation (since it was written thinking we were going to keep the staging dir around).

Reply:
If we want to change the behavior of removing staging dir, it can be done at low cost (I believe).
We can save staging dir clean command to another list attribute. And on close call decide to clean or not this list depending on the recovery status (not sure it is available at that point or if it is a clean and clear status).

Reply:
But I would keep this as a future improvement.